src/tests/itests/xi_itest_connect_error.c

```diff
@@ -423,3 +423,396 @@ void xi_itest_test_valid_flow__call_connect_function_then_disconnect_without_mak
+        expect_value( xi_mock_layer_tls_prev_connect, in_out_state, XI_STATE_OK );
+        expect_value( xi_mock_broker_layer_connect, in_out_state, XI_STATE_OK );
+
+        expect_value( xi_mock_broker_layer_init, in_out_state, XI_STATE_OK );
+        expect_value( xi_mock_broker_layer_connect, in_out_state, XI_STATE_OK );
+
+        /* no problem during CONNECT*/
+        expect_value( xi_mock_broker_layer_push, in_out_state, XI_STATE_OK );
+        expect_value( xi_mock_broker_layer_push, in_out_state, XI_STATE_WRITTEN );
+        expect_value( xi_mock_layer_tls_prev_push, in_out_state, XI_STATE_OK );
+
+        /* CONNECT message arrives at mock broker*/
+        expect_value( xi_mock_broker_layer_pull, in_out_state, XI_STATE_OK );
+        expect_value( xi_mock_broker_layer_pull, recvd_msg_type, XI_MQTT_TYPE_CONNECT );
+
+        /* CONNACK sent*/
+        expect_value( xi_mock_broker_secondary_layer_push, in_out_state, XI_STATE_OK );
+        will_return( xi_mock_broker_secondary_layer_push, CONTROL_CONTINUE );
+        expect_value( xi_mock_broker_layer_push, in_out_state, XI_STATE_WRITTEN );
+
+
+#ifdef XI_CONTROL_TOPIC_ENABLED
+        /* second message (probably SUBSCRIBE on a control topic)*/
+        expect_value( xi_mock_broker_layer_push, in_out_state, XI_STATE_OK );
+        expect_value( xi_mock_broker_layer_push, in_out_state, XI_STATE_WRITTEN );
+        expect_value( xi_mock_layer_tls_prev_push, in_out_state, XI_STATE_OK );
+
+        /* SUBSCRIBE message arrives at mock broker*/
+        expect_value( xi_mock_broker_layer_pull, in_out_state, XI_STATE_OK );
+        expect_value( xi_mock_broker_layer_pull, recvd_msg_type, XI_MQTT_TYPE_SUBSCRIBE );
+
+        expect_string( xi_mock_broker_layer_pull, subscribe_topic_name,
+                       fixture->control_topic_name );
+
+        /* SUBACK sent*/
+        expect_value( xi_mock_broker_secondary_layer_push, in_out_state, XI_STATE_OK );
+        will_return( xi_mock_broker_secondary_layer_push, CONTROL_CONTINUE );
+        expect_value( xi_mock_broker_layer_push, in_out_state, XI_STATE_WRITTEN );
+#endif
+
+        /* DISCONNECT MESSAGE */
+        expect_value( xi_mock_broker_layer_push, in_out_state, XI_STATE_OK );
+        expect_value( xi_mock_layer_tls_prev_push, in_out_state, XI_STATE_OK );
+        expect_value( xi_mock_broker_layer_push, in_out_state, XI_STATE_WRITTEN );
+        expect_value( xi_mock_broker_layer_pull, in_out_state, XI_STATE_OK );
+        expect_value( xi_mock_broker_layer_pull, recvd_msg_type,
+                      XI_MQTT_TYPE_DISCONNECT );
+
+        /* SHUTDOWN */
+        expect_value( xi_mock_broker_layer_close, in_out_state, XI_STATE_OK );
+        expect_value( xi_mock_broker_layer_close, in_out_state, XI_STATE_OK );
+        expect_value( xi_mock_layer_tls_prev_close, in_out_state, XI_STATE_OK );
+        expect_value( xi_mock_layer_tls_prev_close_externally, in_out_state,
```

Comment:
please take a look at xi_itest_sft.c's test cases. It turns of the observation of mock broker layer functions, this means mass reduction of expect_value... functions which are totally unrelated in these test cases since you want to test the is_connected functionality and not the mock broker's behaviour. Especially mock broker's behaviour is tested already in other test cases.
So for doing that look at the xi_itest_sft.c:xi_itest_sft__act function's first lines:
will_return_always( xi_mock_broker_layer__check_expected__LAYER_LEVEL,
                            CONTROL_SKIP_CHECK_EXPECTED );
This turns off observation of layer level functions. mqtt level functions can be turned off too and layer level functions for tls_prev layer. 
With this move you can get rid of the long expect_value block pattern.

Reply:
With this the test cases will become clearer and focused on really what they test.

Reply:
hmm. not exactly sure how to use these functions.  Think I'll need another join.me session. :(